File: backend/app/routes/socios.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required

from ..extensions import db
from ..models.socio import Socio, ESTADOS_PLAN
from ..utils.decorators import roles_required

socios_bp = Blueprint('socios', __name__)


@socios_bp.route('', methods=['GET'])
@jwt_required()
def listar_socios():
    query = Socio.query

    nombre = request.args.get('nombre')
    dni = request.args.get('dni')
    tipo = request.args.get('tipo')
    estado_plan = request.args.get('estado_plan')
    es_socio = request.args.get('es_socio')

    if nombre:
        query = query.filter(Socio.nombre_completo.ilike(f'%{nombre}%'))
    if dni:
        query = query.filter(Socio.dni.ilike(f'%{dni}%'))
    if tipo:
        query = query.filter(Socio.tipo == tipo)
    if estado_plan:
        query = query.filter(Socio.estado_plan == estado_plan)
    if es_socio is not None:
        query = query.filter(Socio.es_socio == (es_socio.strip().lower() not in ('0', 'false', 'no', '')))

    socios = query.order_by(Socio.nombre_completo).all()
    return jsonify([s.to_dict() for s in socios])


@socios_bp.route('/<int:socio_id>', methods=['GET'])
@jwt_required()
def obtener_socio(socio_id):
    socio = Socio.query.get_or_404(socio_id)
    return jsonify(socio.to_dict())


@socios_bp.route('', methods=['POST'])
@roles_required('ADMIN', 'BIBLIOTECARIO')
def crear_socio():
    data = request.get_json(silent=True) or {}
    nombre = (data.get('nombre_completo') or '').strip()
    dni = (data.get('dni') or '').strip()

    if not nombre or not dni:
        return jsonify({'error': 'Nombre completo y DNI son obligatorios'}), 400

    if Socio.query.filter_by(dni=dni).first():
        return jsonify({'error': 'Ya existe una persona registrada con ese DNI'}), 409

    es_socio = data.get('es_socio', True)
    if isinstance(es_socio, str):
        es_socio = es_socio.strip().lower() not in ('0', 'false', 'no', '')

    socio = Socio(
        tipo=data.get('tipo') or 'alumno',
        nombre_completo=nombre,
        dni=dni,
        telefono=data.get('telefono'),
        email=data.get('email'),
        curso=data.get('curso'),
        division=data.get('division'),
        materia=data.get('materia'),
        estado_plan='al_dia',
        es_socio=bool(es_socio),
    )
    db.session.add(socio)
    db.session.commit()
    return jsonify(socio.to_dict()), 201


@socios_bp.route('/<int:socio_id>', methods=['PUT'])
@roles_required('ADMIN', 'BIBLIOTECARIO')
def actualizar_socio(socio_id):
    socio = Socio.query.get_or_404(socio_id)
    data = request.get_json(silent=True) or {}

    campos = ['tipo', 'nombre_completo', 'telefono', 'email', 'curso', 'division', 'materia']
    for campo in campos:
        if campo in data:
            setattr(socio, campo, data[campo])

    if 'estado_plan' in data and data['estado_plan'] in ESTADOS_PLAN:
        socio.estado_plan = data['estado_plan']

    if 'es_socio' in data:
        valor = data['es_socio']
        if isinstance(valor, str):
            valor = valor.strip().lower() not in ('0', 'false', 'no', '')
        socio.es_socio = bool(valor)

    db.session.commit()
    return jsonify(socio.to_dict())


@socios_bp.route('/<int:socio_id>', methods=['DELETE'])
@roles_required('ADMIN')
def eliminar_socio(socio_id):
    socio = Socio.query.get_or_404(socio_id)
    db.session.delete(socio)
    db.session.commit()
    return jsonify({'mensaje': 'Socio eliminado'})

File: server/routes/socios.js
const express = require('express');

const { one, many, run } = require('../db');
const { requireAuth, requireRoles } = require('../middleware/auth');

const router = express.Router();

const ESTADOS_PLAN = ['al_dia', 'suspendido', 'bloqueado'];

function toBool(valor, porDefecto = true) {
  if (valor === undefined || valor === null) return porDefecto;
  if (typeof valor === 'string') {
    return !['0', 'false', 'no', ''].includes(valor.trim().toLowerCase());
  }
  return Boolean(valor);
}

function toDictSocio(s) {
  return {
    id: s.id,
    tipo: s.tipo,
    nombre_completo: s.nombre_completo,
    dni: s.dni,
    telefono: s.telefono,
    email: s.email,
    curso: s.curso,
    division: s.division,
    materia: s.materia,
    estado_plan: s.estado_plan,
    es_socio: Boolean(s.es_socio),
  };
}

router.get('/', requireAuth, async (req, res) => {
  const { nombre, dni, tipo, estado_plan, es_socio } = req.query;

  let sql = 'SELECT * FROM socios WHERE 1=1';
  const params = [];
  if (nombre) { params.push(`%${nombre}%`); sql += ` AND nombre_completo ILIKE $${params.length}`; }
  if (dni) { params.push(`%${dni}%`); sql += ` AND dni ILIKE $${params.length}`; }
  if (tipo) { params.push(tipo); sql += ` AND tipo = $${params.length}`; }
  if (estado_plan) { params.push(estado_plan); sql += ` AND estado_plan = $${params.length}`; }
  if (es_socio !== undefined) { params.push(toBool(es_socio)); sql += ` AND es_socio = $${params.length}`; }
  sql += ' ORDER BY nombre_completo';

  const socios = await many(sql, params);
  res.json(socios.map(toDictSocio));
});

router.get('/:id', requireAuth, async (req, res) => {
  const socio = await one('SELECT * FROM socios WHERE id = $1', [Number(req.params.id)]);
  if (!socio) return res.status(404).json({ error: 'Recurso no encontrado' });
  res.json(toDictSocio(socio));
});

router.post('/', requireRoles('ADMIN', 'BIBLIOTECARIO'), async (req, res) => {
  const data = req.body || {};
  const nombre = (data.nombre_completo || '').trim();
  const dni = (data.dni || '').trim();

  if (!nombre || !dni) {
    return res.status(400).json({ error: 'Nombre completo y DNI son obligatorios' });
  }
  if (await one('SELECT id FROM socios WHERE dni = $1', [dni])) {
    return res.status(409).json({ error: 'Ya existe una persona registrada con ese DNI' });
  }

  const esSocio = toBool(data.es_socio, true);
  const socio = await one(
    `INSERT INTO socios (tipo, nombre_completo, dni, telefono, email, curso, division, materia, estado_plan, es_socio)
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8, 'al_dia', $9) RETURNING *`,
    [
      data.tipo || 'alumno',
      nombre,
      dni,
      data.telefono || null,
      data.email || null,
      data.curso || null,
      data.division || null,
      data.materia || null,
      esSocio,
    ]
  );

  res.status(201).json(toDictSocio(socio));
});

router.put('/:id', requireRoles('ADMIN', 'BIBLIOTECARIO'), async (req, res) => {
  const socioId = Number(req.params.id);
  const socio = await one('SELECT * FROM socios WHERE id = $1', [socioId]);
  if (!socio) return res.status(404).json({ error: 'Recurso no encontrado' });

  const data = req.body || {};
  const campos = ['tipo', 'nombre_completo', 'telefono', 'email', 'curso', 'division', 'materia'];
  const actualizaciones = campos.filter((c) => c in data);
  if (actualizaciones.length) {
    const valores = actualizaciones.map((c) => data[c]);
    const set = actualizaciones.map((c, i) => `${c} = $${i + 1}`).join(', ');
    await run(`UPDATE socios SET ${set} WHERE id = $${valores.length + 1}`, [...valores, socioId]);
  }

  if ('estado_plan' in data && ESTADOS_PLAN.includes(data.estado_plan)) {
    await run('UPDATE socios SET estado_plan = $1 WHERE id = $2', [data.estado_plan, socioId]);
  }

  if ('es_socio' in data) {
    await run('UPDATE socios SET es_socio = $1 WHERE id = $2', [toBool(data.es_socio), socioId]);
  }

  const actualizado = await one('SELECT * FROM socios WHERE id = $1', [socioId]);
  res.json(toDictSocio(actualizado));
});

router.delete('/:id', requireRoles('ADMIN'), async (req, res) => {
  const socioId = Number(req.params.id);
  const socio = await one('SELECT * FROM socios WHERE id = $1', [socioId]);
  if (!socio) return res.status(404).json({ error: 'Recurso no encontrado' });

  await run('DELETE FROM socios WHERE id = $1', [socioId]);
  res.json({ mensaje: 'Socio eliminado' });
});

module.exports = { router, toDictSocio };
